docs/developer/design/authentication.rst

```diff
@@ -0,0 +1,137 @@
+
+- Is support for Dex monitoring and alerting on Grafana
+  something we should think about now?
+
+Abstract
+--------
+
+To be able to identify who a user is and to manage(Create/Edit/Delete)
+these user identities, MetalK8s needs a middleman that acts in this light
+as an Identity Provider.
+
+To accomplish this, we introduce the concept of an OpenID Connect identity
+provider which has as role identity federation in a typical MetalK8s cluster.
+
+
+Scope
+-----
+
+The scope of **Authentication** will be split into 3 main iterations as
+detailed in the implementation section of this document.
+
+Requirements
+------------
+
+User Requirements for Authentication in MetalK8s has been discussed and is
+well documented. For more information, refer to -> [link to Requirements DOC]
+
+
+Design
+------
+
+For authentication in **MetalK8s**, we are going to use Dex, Dex
+connectors/plugins and it's integrated support for other
+External IDP Integration as an Identity Provider.
+
+Design choices
+^^^^^^^^^^^^^^
+
+Dex is chosen as an Identity Provider in MetalK8s based on the above
+requirements enforced by the following reasons:
+
+* Dex's support for multiple plugins greatly increases the potential for
+  integrating with existing user management system such as Active Directory,
+  LDAP, SAML and others.
+* The ability for Dex to issue tokens, revoke tokens, refresh tokens and even
+  control the lifetime of these tokens makes **Dex** suitable since we can
+  force users to re-authenticate when need arises.
+* Dex can be seamlessly deployed in a Kubernetes cluster hence making it
+  easy to handle user creation and modification easily.
+* Dex provides access to a highly customizable UI which is a step closer to
+  good user experience which we advocate for.
+* Dex provides support for multiple backends for storage. This guarantees user
+  account information persistence after upgrades.
+
+Rejected design choices
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Static passwords
+~~~~~~~~~~~~~~~~
+
+This involves adding new users by updating a static file located on every
+control-plane Node and then restarting the API server.
+
+This was rejected since there is no support for a pluggable user management
+system and it is is cumbersome to operate.
+
+X.509 certificates
+~~~~~~~~~~~~~~~~~~
+
+Here, each user owns a signed certificate that is validated by the API
+server.
+
+This approach is not user-friendly meaning each certificate has to be manually
+signed. Also providing certificates for accessing the MetalK8s UI needs
+much more effort.
+
+Keycloak
+~~~~~~~~
+
+
+Implementation Details
+----------------------
+
+Implementation steps(High level)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Iteration 1
+~~~~~~~~~~~
+
+- Deploy Dex in MetalK8s using the official **Dex Charts**
+- Setup Dex to use self-signed certificates
+- Expose Dex on the control plane using Ingress
+- Configure Kubernetes API server to use Dex
+- Provide documentation(Dex + API server integration)
```

Comment:
Why should contain this documentation if DEX + API server integration is done automatically when deploying bootstrap node and expanding control plane ?